group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2643,9 +2652,175 @@ private CoordinatorResult<Void, Record> updateStaticMemberAndRebalance(
+                );
+                return new CoordinatorResult<>(records, appendFuture);
+            }
+
+        } else if (group.isInState(STABLE)) {
+            removePendingSyncMember(group, memberId);
+
+            // If the group is stable, we just return the current assignment
+            GenericGroupMember member = group.member(memberId);
+            responseFuture.complete(new SyncGroupResponseData()
+                .setProtocolType(group.protocolType().orElse(null))
+                .setProtocolName(group.protocolName().orElse(null))
+                .setAssignment(member.assignment())
+                .setErrorCode(Errors.NONE.code()));
+
+        } else if (group.isInState(DEAD)) {
+            throw new IllegalStateException("Reached unexpected condition for Dead group " + groupId);
+        }
+
         return EMPTY_RESULT;
     }
 
+    private Optional<Errors> validateSyncGroup(
+        GenericGroup group,
+        SyncGroupRequestData request
+    ) {
+        if (group.isInState(DEAD)) {
+            // If the group is marked as dead, it means some other thread has just removed the group
+            // from the coordinator metadata; this is likely that the group has migrated to some other
+            // coordinator OR the group is in a transient unstable phase. Let the member retry
+            // finding the correct coordinator and rejoin.
+            return Optional.of(COORDINATOR_NOT_AVAILABLE);
+        } else {
+            Optional<Errors> memberError = validateExistingMember(
+                group,
+                request.memberId(),
+                request.groupInstanceId(),
+                "sync-group"
+            );
+            if (memberError.isPresent()) {
+                return memberError;
+            } else {
+                if (request.generationId() != group.generationId()) {
+                    return Optional.of(Errors.ILLEGAL_GENERATION);
+
+                } else if (request.protocolType() != null &&
+                    group.protocolType().isPresent() &&
+                    !group.protocolType().get().equals(request.protocolType())
+                ) {
+                    return Optional.of(Errors.INCONSISTENT_GROUP_PROTOCOL);
+
+                } else if (request.protocolName() != null &&
+                    group.protocolName().isPresent() &&
+                    !group.protocolName().get().equals(request.protocolName())
+                ) {
+                    return Optional.of(Errors.INCONSISTENT_GROUP_PROTOCOL);
+
+                } else {
+                    return Optional.empty();
+                }
+            }
+        }
+    }
+
+    private void removePendingSyncMember(
+        GenericGroup group,
+        String memberId
+    ) {
+        group.removePendingSyncMember(memberId);
+
+        String syncKey = syncKey(group.groupId());
+
+        if (group.generationId() != group.pendingSyncGenerationId()) {
```

Comment:
Is this required here? I think that this is required when the timeout expires because we want to ensure that the stale timer has no impact. However it does not seem to be required here.

Reply:
So that i understand: let's say that there is a pending sync with an invalid generation id that expired. Since GroupMetadataManager#expirePendingSync already checks and ignores stale generation ids, we don't need to check & cancel it here?

Reply:
Yeah, that's my understanding.